source/connection-monitoring-and-pooling/tests/README.rst

```diff
@@ -0,0 +1,165 @@
+.. role:: javascript(code)
+  :language: javascript
+
+========================================
+Connection Monitoring and Pooling (CMaP)
+========================================
+
+.. contents::
+
+--------
+
+Introduction
+============
+
+The YAML and JSON files in this directory are platform-independent tests that
+drivers can use to prove their conformance to the Connection Monitoring and Pooling (CPAM) Spec.
+
+Several prose tests, which are not easily expressed in YAML, are also presented
+in this file. Those tests will need to be manually implemented by each driver.
+
+Common Test Format
+================
+
+Each YAML file has the following keys:
+
+- ``version``: A version number indicating the expected format of the spec tests (current version = 1)
+- ``style``: A string indicating what style of tests this file contains. Currently ``unit`` is the only valid value
+- ``tests``: An array of tests that are to be run independently of each other
+
+Unit Test Format:
+=================
+
+All Unit Tests have some of the following fields:
+
+- ``description``: A text description of what the test is meant to assert
+- ``operations``: A list of operations to perform. Each operation has some of the following fields:
+
+  - ``command``: A string describing which command to issue.
+  - ``args``: An array of arguments to pass to the command
+  - ``thread``: The name of the thread in which to run this command. If not specified, runs in the default thread
+  - ``object``: For some command, specifies the object on which to run the command
+  - ``returnTo``: For some commands, specifies what to name the return value of the command
+
+- ``error``: Indicates that the main thread is expected to error during this test. Subfields (like ``message``) indicate what error is expected
+- ``events``: An array of all connection monitoring events expected to occur while running ``operations``
```

Comment:
I think documentation of the subfields of each `event` would be useful so that implementers don't have to read through all of the tests to determine what the format of an `event` is.

Reply:
The `event` here maps to whatever fields are specified for an event in the specification. I can, in the final version, link to the `event` section of the spec, but I'd prefer to not duplicate that information in multiple places. It would make adding fields to any single event require a change here. And I am hoping that implementers actually read the spec itself to learn the format of the events rather than just the test runner description.